src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs

```diff
@@ -96,15 +98,127 @@ protected ServiceEndpoint[] GetValuableEndpoints(IEnumerable<ServiceEndpoint> en
             return groupedEndpoints.ToArray();
         }
 
+        protected async Task AddServiceEndpointsAsync(IReadOnlyList<ServiceEndpoint> endpoints)
+        {
+            if (endpoints.Count > 0)
+            {
+                var hubEndpoints = CreateHubServiceEndpoints(endpoints);
+
+                await Task.WhenAll(hubEndpoints.Select(e => AddHubServiceEndpointAsync(e)));
+
+                // TODO: update local store for negotiation
+            }
+        }
+
+        protected async Task RemoveServiceEndpointsAsync(IReadOnlyList<ServiceEndpoint> endpoints)
+        {
+            if (endpoints.Count > 0)
+            {
+                var hubEndpoints = CreateHubServiceEndpoints(endpoints);
+
+                // TODO: update local store for negotiation
+
+                await Task.WhenAll(hubEndpoints.Select(e => RemoveHubServiceEndpointAsync(e)));
+            }
+        }
+
+        protected Task RenameSerivceEndpoints(IReadOnlyList<ServiceEndpoint> endpoints)
+        {
+            if (endpoints.Count > 0)
+            {
+                var hubEndpoints = CreateHubServiceEndpoints(endpoints);
+
+                // TODO: update local store for negotiation
+
+                return Task.WhenAll(hubEndpoints.Select(e => RenameHubServiceEndpoint(e)));
+            }
+            return Task.CompletedTask;
+        }
+
+        private HubServiceEndpoint CreateHubServiceEndpoint(string hub, ServiceEndpoint endpoint)
+        {
+            var provider = GetEndpointProvider(endpoint);
+            return new HubServiceEndpoint(hub, provider, endpoint);
+        }
+
+        private IReadOnlyList<HubServiceEndpoint> CreateHubServiceEndpoints(string hub, IEnumerable<ServiceEndpoint> endpoints)
+        {
+            return endpoints.Select(e => CreateHubServiceEndpoint(hub, e)).ToList();
+        }
+
+        private IReadOnlyList<HubServiceEndpoint> CreateHubServiceEndpoints(IEnumerable<ServiceEndpoint> endpoints)
+        {
+            var hubEndpoints = new List<HubServiceEndpoint>();
+            var hubs = _endpointsPerHub.Keys;
+            foreach (var hub in hubs)
+            {
+                hubEndpoints.AddRange(CreateHubServiceEndpoints(hub, endpoints));
+            }
+            return hubEndpoints;
+        }
+
+        private async Task AddHubServiceEndpointAsync(HubServiceEndpoint endpoint)
+        {
+            Log.StartAddingEndpoint(_logger, endpoint.Endpoint, endpoint.Name);
+            endpoint.Ready = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            OnAdd?.Invoke(endpoint);
+            // Wait for new endpoint turn Ready
+            await endpoint.Ready.Task;
```

Comment:
who handles if it hung?

Reply:
Suppose in Container side will set true finally. But to ensure unblock, we may add timeout check here as well. But in this case I think it probably means connection may not even starts, and shall we block add negotiation as well?